File: services/job-service/src/cache/redisCache.js
'use strict'

const crypto = require('crypto')
const { createClient } = require('redis')

let client
let ready = false
let connecting = false

const SEARCH_KEY_SET = 'cache:jobs:search:keys'

function isEnabled () {
  return Boolean(process.env.REDIS_URL || process.env.REDIS_HOST)
}

function ttlSearchSeconds () {
  const n = Number(process.env.REDIS_TTL_SEARCH_SECONDS)
  return Number.isInteger(n) && n > 0 ? n : 60
}

function ttlGetSeconds () {
  const n = Number(process.env.REDIS_TTL_GET_SECONDS)
  return Number.isInteger(n) && n > 0 ? n : 10
}

function redisUrl () {
  if (process.env.REDIS_URL) return process.env.REDIS_URL
  const host = process.env.REDIS_HOST || '127.0.0.1'
  const port = Number(process.env.REDIS_PORT) || 6379
  return `redis://${host}:${port}`
}

async function getClient () {
  if (!isEnabled()) return null
  if (client && ready) return client
  if (connecting) return null

  connecting = true
  try {
    if (!client) {
      client = createClient({ url: redisUrl() })
      client.on('error', () => {
        ready = false
      })
    }
    if (!client.isOpen) await client.connect()
    ready = true
    return client
  } catch {
    ready = false
    return null
  } finally {
    connecting = false
  }
}

function hashPayload (payload) {
  return crypto.createHash('sha256').update(JSON.stringify(payload)).digest('hex')
}

function searchKey (payload) {
  return `cache:jobs:search:${hashPayload(payload)}`
}

function getJobKey (jobId) {
  return `cache:jobs:get:${jobId}`
}

async function getJson (key) {
  const c = await getClient()
  if (!c) return null
  try {
    const raw = await c.get(key)
    if (!raw) return null
    return JSON.parse(raw)
  } catch {
    return null
  }
}

async function setJson (key, value, ttlSeconds) {
  const c = await getClient()
  if (!c) return false
  try {
    await c.set(key, JSON.stringify(value), { EX: ttlSeconds })
    return true
  } catch {
    return false
  }
}

async function getSearchCache (payload) {
  return getJson(searchKey(payload))
}

async function setSearchCache (payload, result) {
  const c = await getClient()
  if (!c) return false
  const key = searchKey(payload)
  try {
    await c.set(key, JSON.stringify(result), { EX: ttlSearchSeconds() })
    await c.sAdd(SEARCH_KEY_SET, key)
    return true
  } catch {
    return false
  }
}

async function getJobCache (jobId) {
  return getJson(getJobKey(jobId))
}

async function setJobCache (jobId, result) {
  return setJson(getJobKey(jobId), result, ttlGetSeconds())
}

async function invalidateJobCache (jobId) {
  const c = await getClient()
  if (!c) return
  try {
    await c.del(getJobKey(jobId))
  } catch {
    // ignore
  }
}

async function invalidateAllSearchCache () {
  const c = await getClient()
  if (!c) return
  try {
    const keys = await c.sMembers(SEARCH_KEY_SET)
    if (keys.length) await c.del(keys)
    await c.del(SEARCH_KEY_SET)
  } catch {
    // ignore
  }
}

async function disconnectCache () {
  if (!client) return
  try {
    if (client.isOpen) await client.quit()
  } catch {
    // ignore
  } finally {
    ready = false
    client = null
  }
}

module.exports = {
  getSearchCache,
  setSearchCache,
  getJobCache,
  setJobCache,
  invalidateJobCache,
  invalidateAllSearchCache,
  disconnectCache
}

File: services/job-service/backend/app/cache/redis_cache.py
import hashlib
import json
import threading

import redis

from app.core.config import settings

SEARCH_KEY_SET = "cache:jobs:search:keys"
_lock = threading.Lock()
_client: redis.Redis | None = None


def _redis_enabled() -> bool:
    return bool(settings.redis_url or settings.redis_host)


def _redis_url() -> str:
    if settings.redis_url:
        return settings.redis_url
    return f"redis://{settings.redis_host or '127.0.0.1'}:{settings.redis_port}"


def _get_client() -> redis.Redis | None:
    global _client
    if not _redis_enabled():
        return None
    with _lock:
        if _client is None:
            try:
                _client = redis.Redis.from_url(_redis_url(), decode_responses=True)
                _client.ping()
            except redis.RedisError:
                _client = None
        return _client


def _ttl_search() -> int:
    n = settings.redis_ttl_search_seconds
    return n if isinstance(n, int) and n > 0 else 60


def _ttl_get() -> int:
    n = settings.redis_ttl_get_seconds
    return n if isinstance(n, int) and n > 0 else 10


def _hash_payload(payload: dict) -> str:
    return hashlib.sha256(json.dumps(payload, sort_keys=True, default=str).encode()).hexdigest()


def search_key(payload: dict) -> str:
    return f"cache:jobs:search:{_hash_payload(payload)}"


def get_job_key(job_id: str) -> str:
    return f"cache:jobs:get:{job_id}"


def get_search_cache(payload: dict):
    c = _get_client()
    if not c:
        return None
    try:
        raw = c.get(search_key(payload))
        return json.loads(raw) if raw else None
    except (redis.RedisError, json.JSONDecodeError, TypeError):
        return None


def set_search_cache(payload: dict, result: dict) -> bool:
    c = _get_client()
    if not c:
        return False
    key = search_key(payload)
    try:
        c.set(key, json.dumps(result, default=str), ex=_ttl_search())
        c.sadd(SEARCH_KEY_SET, key)
        return True
    except redis.RedisError:
        return False


def get_job_cache(job_id: str):
    c = _get_client()
    if not c:
        return None
    try:
        raw = c.get(get_job_key(job_id))
        return json.loads(raw) if raw else None
    except (redis.RedisError, json.JSONDecodeError, TypeError):
        return None


def set_job_cache(job_id: str, result: dict) -> bool:
    c = _get_client()
    if not c:
        return False
    try:
        c.set(get_job_key(job_id), json.dumps(result, default=str), ex=_ttl_get())
        return True
    except redis.RedisError:
        return False


def invalidate_job_cache(job_id: str) -> None:
    c = _get_client()
    if not c:
        return
    try:
        c.delete(get_job_key(job_id))
    except redis.RedisError:
        pass


def invalidate_all_search_cache() -> None:
    c = _get_client()
    if not c:
        return
    try:
        keys = list(c.smembers(SEARCH_KEY_SET))
        if keys:
            c.delete(*keys)
        c.delete(SEARCH_KEY_SET)
    except redis.RedisError:
        pass


def disconnect_cache() -> None:
    global _client
    with _lock:
        if _client:
            try:
                _client.close()
            except redis.RedisError:
                pass
            _client = None
